12-watchtower-API.md

```diff
@@ -22,227 +22,320 @@ The scope of this bolt does not include:
-	* MUST fail the connection
-* upon receiving a `modes` that does not contain any of its own accepted modes:
-	* MUST fail the connection
-* upon receiving a `qos` that does not contain any of its own accepted QoS:
-	* MUST fail the connection
+The receiving node SHOULD reject the appointment if:
+
+* `end_block` is too far away in the future.
+
+The receiving node MUST: 
+
+* truncate the remainder of the package to `qos_len`.
+* if `qos_len` is not 0:
+	* process `qos_data` according to [Quality of Service data](#quality-of-service-data).
+
+The receiving node MAY accept the appointment otherwise.
 
 #### Rationale
 
-The client is always the one in charge of establishing the connection. His `wt_init` message informs the server of what type of service he is requesting. The server should agree with the client if the implement the same methods. Otherwise the connection should be failed.
+WatchTowers may work in different modes (e.g. altruistic vs non-altruistic). In some of those modes, proof of payment may be required for the WatchTower to perform provide the service. `auth_token` is used to decide wether the user is entitled to use the service or not in the cases it may be required. Notice that the tokens do not need to be linked to any kind of identity but confirm that the payment has been performed.
 
-QoS is an optional field. Including the field in the `wt_init` message signals that the sender is requiring that specific QoS. As for accountability, it aims for giving non-repudiable proof of the agreement to both the client and the server.
+The transaction `locator` can be deterministically computed by both the client and the server. Locators of wrong size are therefore invalid.
 
-The transport protocol to be used is purposely ommited. Piggybacking on top of the Lightning transport protocol as well other approaches such as interfaces over HTTP can be used to establish the connection.
+`start_block` can be either one block ahead or behind the tower tip due to network delays. A tower must not accept appointments arbitrarily ahead or behind the current tip since it could ease DoS vectors. A `start_block` long behind would force the tower to rescan block data for those appointments instead of watching block by block. On the other hand, a `start_time` long ahead would imply storing information way before it being needed.
 
-## Sending / receiving appointments
+Regarding the `end_block`, too far away is a subjective concept. The further away a tower accepts appointment ends, the higher the potential storage requirements may be, and the easier (and cheaper) would it be to DoS.
 
-Once both client and server have agreed on common modes of operation, the client can start sending appointments to the server.
+The `encrypted_blob` should have been encrypted using `cipher`. Block ciphers have a size multiple of the block length, which depends on the key size. Therefore the `encrypted_blob` have to be at least as big as:
```

Comment:
shouldn't "should" be "must"?

Reply:
Well, the tower assumes so, but a malicious user may have done differently. So the tower does not know but can do some checks narrow this down

Reply:
and the cheaper it may be to perform a denial of service*. 

the 'encrypted_blob' should be encrypted 

(lets stick to present tense).

Reply:
`start_block` can be either one block ahead or behind the tower tip to take into account network delays. It is advisable to avoid breaching this range as it opens potential DoS vectors.

This makes me wonder. Do we need to re-scan blocks for older jobs? What if a user has had a channel open for 1 week and then wants to send us all the justice transactions in a big batch? Maybe one thing we need to do is keep around a mapping of all spent utxo?

Reply:
But I guess then the start_block would be the latest block, even in the case of a batch? Maybe another way to put it; is that we are using the start_block to check freshness of the job?

Reply:
We should not reescan anything from old blocks. If a customer wants to hire a tower to watch for an old state it can do it from that point on, but accepting jobs that started in the past make no sense. If we do then:

- We have to reescan what's costly and open even more DoS
- We may automatically fail a job since a breach with no justice may be already in the blockchain.

So accepting old states (in batches or one by one) is not a problem, but Watching in the past doesn't make sense.

Reply:
ya. so what we should probably do is require it to be for the next block onwards? i.e. start 1-2 blocks ahead of the current? That way we do not have to do any scanning?

We should also put in the FIXME: How to deal with batch jobs. i.e. the client is selecting a new watchtower and has to send ALL revoked sates up. (a potentially big task). That should argubly be a new BOLT / a new tag 'batch-job'. We can assume the customer has a fresh channel with us (i.e. they just reset channel onchain before using us).

Reply:
Resolved offlline

Reply:
does the resolved include the fixme about batch jobs?

Reply:
By design jobs are not linked between each other. We can have discussion about this and having an extension that allows it. I can add it in the fixme section for now.

Reply:
I'll add an item in the fixmes to discuss about that / have a proposal for later iterations of the draft.